File: cloud/blockstore/public/sdk/python/client/session.py
import copy
import enum
import logging
import threading

from concurrent import futures
from datetime import datetime
from logging import Logger
from typing import Callable

import cloud.blockstore.public.sdk.python.protos as protos

from .client import SessionInfo
from .error import ClientError, _handle_errors
from .error_codes import EResult
from .scheduler import Scheduler


class _EMountState(enum.Enum):
    UNINITIALIZED = 0,
    MOUNT_COMPLETED = 1,
    MOUNT_REQUESTED = 2,
    MOUNT_IN_PROGRESS = 3,
    UNMOUNT_IN_PROGRESS = 4


class Session(object):

    def __init__(
            self,
            client,
            disk_id: str,
            mount_token: str,
            access_mode: protos.EVolumeAccessMode = protos.EVolumeAccessMode.Value("VOLUME_ACCESS_READ_WRITE"),
            mount_mode: protos.EVolumeAccessMode = protos.EVolumeMountMode.Value("VOLUME_MOUNT_LOCAL"),
            throttling_disabled: bool = False,
            mount_seq_number: int = 0,
            log: Logger | None = None,
            scheduler : Scheduler | None = None):

        self.__client = client
        self.__disk_id = disk_id
        self.__mount_token = mount_token
        self.__access_mode = access_mode
        self.__mount_mode = mount_mode
        self.__throttling_disabled = throttling_disabled
        self.__mount_seq_number = mount_seq_number

        self.__mount_lock = threading.Lock()

        self.__state = _EMountState.UNINITIALIZED
        self.__volume: protos.TVolume | None = None
        self.__session_info: SessionInfo | None = None

        self.__mount_headers = None

        if scheduler is None:
            self.__scheduler = Scheduler()
        else:
            self.__scheduler = scheduler

        if log is not None:
            self.log = log
        else:
            self.log = logging.getLogger("session")

    @property
    def disk_id(self) -> str:
        return self.__disk_id

    @property
    def access_mode(self) -> protos.EVolumeAccessMode:
        return self.__access_mode

    @property
    def mount_mode(self) -> protos.EVolumeMountMode:
        return self.__mount_mode

    @property
    def throttling_disabled(self) -> bool:
        return self.__throttling_disabled

    @property
    def client(self):
        return self.__client

    @property
    def volume(self) -> protos.TVolume:
        with self.__mount_lock:
            if self.__volume is None:
                return None
            return copy.deepcopy(self.__volume)

    @property
    def info(self) -> SessionInfo | None:
        with self.__mount_lock:
            if self.__session_info is None:
                return None
            return copy.deepcopy(self.__session_info)

    @_handle_errors
    def mount_volume_async(
            self,
            idempotence_id: str | None = None,
            timestamp: datetime | None = None,
            trace_id: str | None = None,
            request_timeout: int | None = None,
            encryption_spec: protos.TEncryptionSpec | None = None) -> futures.Future:

        result = futures.Future()

        with self.__mount_lock:
            if self.__state == _EMountState.MOUNT_IN_PROGRESS:
                error = ClientError(
                    EResult.E_INVALID_STATE.value,
                    "Volume is already being mounted")
                result.set_exception(error)
                return result

            if self.__state == _EMountState.UNMOUNT_IN_PROGRESS:
                error = ClientError(
                    EResult.E_INVALID_STATE.value,
                    "Volume is being unmounted")
                result.set_exception(error)
                return result

            self.__mount_headers = protos.THeaders(
                IdempotenceId=idempotence_id,
                TraceId=trace_id,
                RequestTimeout=request_timeout,
            )

            self.__state = _EMountState.MOUNT_IN_PROGRESS

        self.log.debug("Submit MountVolume request")

        future = self.__client.mount_volume_async(
            self.__disk_id,
            self.__mount_token,
            self.__access_mode,
            self.__mount_mode,
            self.__throttling_disabled,
            self.__mount_seq_number,
            idempotence_id,
            timestamp,
            trace_id,
            request_timeout,
            encryption_spec)

        def process_mount_response(f):
            with self.__mount_lock:
                try:
                    r = f.result()
                    self._process_mount_response(r["Volume"], r["SessionInfo"])
                    result.set_result(r)
                except ClientError as e:
                    self.log.error("MountVolume request failed with error: %s", e)
                    self.__state = _EMountState.MOUNT_REQUESTED
                    result.set_exception(e)

        future.add_done_callback(process_mount_response)
        return result

    @_handle_errors
    def mount_volume(
            self,
            idempotence_id: str | None = None,
            timestamp: datetime | None = None,
            trace_id: str | None = None,
            request_timeout: int | None = None,
            encryption_spec: protos.TEncryptionSpec | None = None) -> dict:

        return self.mount_volume_async(
            idempotence_id,
            timestamp,
            trace_id,
            request_timeout,
            encryption_spec).result()

    @_handle_errors
    def unmount_volume_async(
            self,
            idempotence_id: str | None = None,
            timestamp: datetime | None = None,
            trace_id: str | None = None,
            request_timeout: int | None = None) -> futures.Future:

        session_id = ""
        result = futures.Future()

        with self.__mount_lock:
            if self.__state == _EMountState.UNINITIALIZED:
                error = ClientError(
                    EResult.E_INVALID_STATE.value,
                    "Volume is not mounted")
                result.set_exception(error)
                return result

            if self.__state == _EMountState.UNMOUNT_IN_PROGRESS:
                error = ClientError(
                    EResult.E_INVALID_STATE.value,
                    "Volume is already being unmounted")
                result.set_exception(error)
                return result

            if self.__state == _EMountState.MOUNT_IN_PROGRESS:
                error = ClientError(
                    EResult.E_INVALID_STATE.value,
                    "Volume is being mounted")
                result.set_exception(error)
                return result

            self.__state = _EMountState.UNMOUNT_IN_PROGRESS

            if self.__session_info is not None:
                session_id = self.__session_info.session_id

        self.log.debug("Submit UnmountVolume request")

        future = self.__client.unmount_volume_async(
            self.__disk_id,
            session_id,
            idempotence_id,
            timestamp,
            trace_id,
            request_timeout)

        def process_response(f):
            try:
                self.log.debug("Complete UnmountVolume request")
                response = f.result()
                with self.__mount_lock:
                    self._reset_state()
                result.set_result(response)
            except ClientError as e:
                self.log.error("UnmountVolume request failed with error: %s", e)
                with self.__mount_lock:
                    self.__state = _EMountState.MOUNT_COMPLETED
                result.set_exception(e)

        future.add_done_callback(process_response)
        return result

    @_handle_errors
    def unmount_volume(
            self,
            idempotence_id: str | None = None,
            timestamp: datetime | None = None,
            trace_id: str | None = None,
            request_timeout: int | None = None) -> dict:

        return self.unmount_volume_async(
            idempotence_id,
            timestamp,
            trace_id,
            request_timeout).result()

    @_handle_errors
    def read_blocks_async(
            self,
            start_index: int,
            blocks_count: int,
            checkpoint_id: str,
            idempotence_id: str | None = None,
            timestamp: datetime | None = None,
            trace_id: str | None = None,
            request_timeout: int | None = None) -> futures.Future:

        def read_blocks_impl(session_id):
            return self.__client.read_blocks_async(
                self.__disk_id,
                start_index,
                blocks_count,
                checkpoint_id,
                session_id,
                idempotence_id,
                timestamp,
                trace_id,
                request_timeout)

        future = futures.Future()
        self._process_request(future, read_blocks_impl, "ReadBlocks")
        return future

    @_handle_errors
    def read_blocks(
            self,
            start_index: int,
            blocks_count: int,
            checkpoint_id: str,
            idempotence_id: str | None = None,
            timestamp: datetime | None = None,
            trace_id: str | None = None,
            request_timeout: int | None = None) -> list[bytes]:

        return self.read_blocks_async(
            start_index,
            blocks_count,
            checkpoint_id,
            idempotence_id,
            timestamp,
            trace_id,
            request_timeout).result()

    @_handle_errors
    def write_blocks_async(
            self,
            start_index: int,
            blocks: bytes,
            idempotence_id: str | None = None,
            timestamp: datetime | None = None,
            trace_id: str | None = None,
            request_timeout: int | None = None) -> futures.Future:

        def write_blocks_impl(session_id):
            return self.__client.write_blocks_async(
                self.__disk_id,
                start_index,
                blocks,
                session_id,
                idempotence_id,
                timestamp,
                trace_id,
                request_timeout)

        future = futures.Future()
        self._process_request(future, write_blocks_impl, "WriteBlocks")
        return future

    @_handle_errors
    def write_blocks(
            self,
            start_index: int,
            blocks: bytes,
            idempotence_id: str | None = None,
            timestamp: datetime | None = None,
            trace_id: str | None = None,
            request_timeout: int | None = None):

        return self.write_blocks_async(
            start_index,
            blocks,
            idempotence_id,
            timestamp,
            trace_id,
            request_timeout).result()

    @_handle_errors
    def zero_blocks_async(
            self,
            start_index: int,
            blocks_count: int,
            idempotence_id: str | None = None,
            timestamp: datetime | None = None,
            trace_id: str | None = None,
            request_timeout: int | None = None) -> futures.Future:

        def zero_blocks_impl(session_id):
            return self.__client.zero_blocks_async(
                self.__disk_id,
                start_index,
                blocks_count,
                session_id,
                idempotence_id,
                timestamp,
                trace_id,
                request_timeout)

        future = futures.Future()
        self._process_request(future, zero_blocks_impl, "ZeroBlocks")
        return future

    @_handle_errors
    def zero_blocks(
            self,
            start_index: int,
            blocks_count: int,
            idempotence_id: str | None = None,
            timestamp: datetime | None = None,
            trace_id: str | None = None,
            request_timeout: int | None = None):

        return self.zero_blocks_async(
            start_index,
            blocks_count,
            idempotence_id,
            timestamp,
            trace_id,
            request_timeout).result()

    @_handle_errors
    def _process_mount_response(self, volume: protos.TVolume, session_info: SessionInfo) -> None:
        if self.__volume is not None:
            # validate volume geometry
            if self.__volume.BlockSize != volume.BlockSize:
                raise ClientError(EResult.E_FAIL.value, "Volume geometry changed")

        self.log.debug("Complete MountVolume request")

        self.__volume = volume
        self.__session_info = session_info
        self.__state = _EMountState.MOUNT_COMPLETED

        self.__scheduler.cancel_all()

        if session_info.inactive_clients_timeout != 0:
            def remount():
                self._remount_volume()

            self.__scheduler.schedule(
                session_info.inactive_clients_timeout,
                remount)

    @_handle_errors
    def _remount_volume(self) -> futures.Future | None:
        unmounted = False
        idempotence_id = None
        trace_id = None
        request_timeout = None
        with self.__mount_lock:
            unmounted = (self.__state == _EMountState.UNINITIALIZED)
            idempotence_id = self.__mount_headers.IdempotenceId
            trace_id = self.__mount_headers.TraceId
            request_timeout = self.__mount_headers.RequestTimeout

        if unmounted:
            return None

        self.log.debug("Remount volume")

        future = self.__client.mount_volume_async(
            self.__disk_id,
            self.__mount_token,
            self.__access_mode,
            self.__mount_mode,
            self.__throttling_disabled,
            self.__mount_seq_number,
            idempotence_id=idempotence_id,
            trace_id=trace_id,
            request_timeout=request_timeout)

        def process_mount_response(f):
            with self.__mount_lock:
                try:
                    result = f.result()
                    self._process_mount_response(result["Volume"], result["SessionInfo"])
                except ClientError as e:
                    self.log.error("MountVolume request failed with error: %s", e)
                    self._reset_state()
                    self.__state = _EMountState.MOUNT_REQUESTED

        future.add_done_callback(process_mount_response)
        return future

    def _reset_state(self) -> None:
        self.__state = _EMountState.UNINITIALIZED
        self.__volume = None
        self.__session_info = None
        self.__scheduler.cancel_all()

    @_handle_errors
    def _process_request(self, result: futures.Future, call: Callable[[str], futures.Future], request_name: str) -> None:
        session_id_future = self._ensure_volume_mounted()

        def process_session_id_ready(f):
            session_id = ""
            try:
                session_id = f.result()
            except ClientError as e:
                result.set_exception(e)
                return

            self.log.debug("Submit %s request", request_name)
            call_future = call(session_id)

            def process_call_ready(f):
                try:
                    response = f.result()
                    self.log.debug("Complete %s request", request_name)
                    result.set_result(response)
                except ClientError as e:
                    if e.code != EResult.E_INVALID_SESSION.value:
                        self.log.debug("%s request failed: %s", request_name, e)
                        result.set_exception(e)
                        return

                    with self.__mount_lock:
                        self.__state = _EMountState.MOUNT_REQUESTED

                    self._process_request(result, call, request_name)
            call_future.add_done_callback(process_call_ready)

        session_id_future.add_done_callback(process_session_id_ready)

    @_handle_errors
    def _ensure_volume_mounted(self) -> futures.Future:
        session_id = ""
        needs_remount = False

        result = futures.Future()
        with self.__mount_lock:
            if self.__state == _EMountState.UNINITIALIZED:
                error = ClientError(
                    EResult.E_INVALID_STATE.value,
                    "Volume is not mounted")
                self.log.debug(error)
                result.set_exception(error)
                return result

            needs_remount = (self.__state == _EMountState.MOUNT_REQUESTED)
            if self.__session_info is not None:
                session_id = self.__session_info.session_id

        if not needs_remount:
            result.set_result(session_id)
            return result

        future = self._remount_volume()

        def process_mount_response(f):
            exception = f.exception()
            if exception is not None:
                result.set_exception(exception)
            else:
                r = f.result()
                result.set_result(r["SessionInfo"].session_id)
        future.add_done_callback(process_mount_response)

        return result

File: cloud/blockstore/public/sdk/go/client/session.go
package client

import (
	"sync"
	"time"

	"github.com/golang/protobuf/proto"
	protos "github.com/ydb-platform/nbs/cloud/blockstore/public/api/protos"
	"golang.org/x/net/context"
)

////////////////////////////////////////////////////////////////////////////////

type mountState uint32

const (
	stateUninitialized mountState = iota
	stateMountCompleted
	stateMountRequested
)

////////////////////////////////////////////////////////////////////////////////

type remountCtx struct {
	baseContext context.Context
}

func (*remountCtx) Deadline() (deadline time.Time, ok bool) {
	return
}

func (*remountCtx) Done() <-chan struct{} {
	return nil
}

func (*remountCtx) Err() error {
	return nil
}

func (r *remountCtx) Value(key interface{}) interface{} {
	return r.baseContext.Value(key)
}

func (*remountCtx) String() string {
	return "remountCtx"
}

////////////////////////////////////////////////////////////////////////////////

type remounter struct {
	Ticker     ticker
	Done       chan bool
	remountCtx *remountCtx
}

////////////////////////////////////////////////////////////////////////////////

type Session struct {
	client Client
	log    Log

	mountLock sync.Mutex
	state     mountState
	mountOpts MountVolumeOpts
	volume    *protos.TVolume
	info      SessionInfo
	closed    bool

	mountHeaders protos.THeaders

	makeTicker tickerFactory
	remounter  *remounter
}

func (s *Session) Volume() *protos.TVolume {
	s.mountLock.Lock()
	defer s.mountLock.Unlock()

	return proto.Clone(s.volume).(*protos.TVolume)
}

func (s *Session) Info() SessionInfo {
	s.mountLock.Lock()
	defer s.mountLock.Unlock()

	return s.info
}

func (s *Session) MountVolume(
	ctx context.Context,
	diskId string,
	opts *MountVolumeOpts,
) error {

	s.fetchMountHeaders(ctx)

	if logger := s.log.Logger(LOG_DEBUG); logger != nil {
		logger.Printf(
			ctx,
			"Submit MountVolume for %s, client_id: %v, opts: %v",
			diskId,
			s.mountHeaders.ClientId,
			opts,
		)
	}

	s.mountLock.Lock()
	defer s.mountLock.Unlock()

	if s.closed {
		return &ClientError{
			Code:    E_INVALID_STATE,
			Message: "Session is already closed",
		}
	}

	volume, info, err := s.client.MountVolume(ctx, diskId, opts)

	err = s.processMountResponse(ctx, diskId, volume, info, err)
	if err != nil {
		return err
	}

	if opts != nil {
		s.mountOpts = *opts
	}

	s.volume = volume
	return nil
}

func (s *Session) UnmountVolume(ctx context.Context) error {
	s.mountLock.Lock()
	defer s.mountLock.Unlock()

	if s.closed {
		return &ClientError{
			Code:    E_INVALID_STATE,
			Message: "Session is already closed",
		}
	}

	if s.state == stateUninitialized {
		return &ClientError{
			Code:    E_INVALID_STATE,
			Message: "Volume is not mounted",
		}
	}

	if logger := s.log.Logger(LOG_DEBUG); logger != nil {
		logger.Printf(
			ctx,
			"Submit UnmountVolume for %s, client_id: %v",
			s.volume.DiskId,
			s.mountHeaders.ClientId,
		)
	}

	err := s.client.UnmountVolume(ctx, s.volume.DiskId, s.info.SessionId)
	if err != nil {
		return err
	}

	s.stopRemounter()
	s.resetState()
	return nil
}

func (s *Session) ReadBlocks(
	ctx context.Context,
	startIndex uint64,
	blocksCount uint32,
	checkpointId string,
) ([][]byte, error) {

	var res [][]byte
	err := s.processRequest(
		ctx,
		func(ctx context.Context, sessionId string, diskId string) error {
			var err error
			res, err = s.client.ReadBlocks(
				ctx,
				diskId,
				startIndex,
				blocksCount,
				checkpointId,
				sessionId)
			return err
		})

	return res, err
}

func (s *Session) WriteBlocks(
	ctx context.Context,
	startIndex uint64,
	blocks [][]byte,
) error {

	return s.processRequest(
		ctx,
		func(ctx context.Context, sessionId string, diskId string) error {
			return s.client.WriteBlocks(
				ctx,
				diskId,
				startIndex,
				blocks,
				sessionId)
		})
}

func (s *Session) ZeroBlocks(
	ctx context.Context,
	startIndex uint64,
	blocksCount uint32,
) error {

	return s.processRequest(
		ctx,
		func(ctx context.Context, sessionId string, diskId string) error {
			return s.client.ZeroBlocks(
				ctx,
				diskId,
				startIndex,
				blocksCount,
				sessionId)
		})
}

func (s *Session) Close() {
	s.mountLock.Lock()
	defer s.mountLock.Unlock()

	s.closed = true
	s.stopRemounter()
	s.resetState()
}

func (s *Session) resetState() {
	s.mountOpts = MountVolumeOpts{}
	s.volume = &protos.TVolume{
		DiskId: s.volume.DiskId,
	}
	s.info = SessionInfo{}
	s.state = stateUninitialized
}

func (s *Session) processRequest(
	ctx context.Context,
	call func(ctx context.Context, sessionId string, diskid string) error,
) error {

	for {
		sessionId, diskId, err := s.ensureVolumeMounted(ctx)
		if err != nil {
			return err
		}

		err = call(ctx, sessionId, diskId)

		if GetClientCode(err) != E_INVALID_SESSION {
			return err
		}

		s.mountLock.Lock()
		s.state = stateMountRequested
		s.mountLock.Unlock()
	}
}

func (s *Session) ensureVolumeMounted(ctx context.Context) (string, string, error) {
	s.mountLock.Lock()
	defer s.mountLock.Unlock()

	if s.closed {
		return "", "", &ClientError{
			Code:    E_INVALID_STATE,
			Message: "Session is already closed",
		}
	}

	if s.state == stateUninitialized {
		return "", "", &ClientError{
			Code:    E_INVALID_STATE,
			Message: "Volume is not mounted",
		}
	}

	var err error
	sessionId := s.info.SessionId
	diskId := s.volume.DiskId

	if s.state == stateMountRequested {
		if logger := s.log.Logger(LOG_WARN); logger != nil {
			logger.Printf(
				ctx,
				"Force remount volume %s, client_id: %v",
				diskId,
				s.mountHeaders.ClientId,
			)
		}

		sessionId, err = s.remountVolume(ctx)
		if err != nil {
			return "", "", err
		}
	}

	return sessionId, diskId, nil
}

func (s *Session) remountVolume(ctx context.Context) (string, error) {

	ctx = s.setupMountHeaders(ctx)
	diskId := s.volume.DiskId

	if logger := s.log.Logger(LOG_DEBUG); logger != nil {
		logger.Printf(
			ctx,
			"Remount volume %s, client_id: %v",
			diskId,
			s.mountHeaders.ClientId,
		)
	}

	volume, info, err := s.client.MountVolume(
		ctx,
		diskId,
		&s.mountOpts)

	err = s.processMountResponse(ctx, diskId, volume, info, err)
	if err != nil {
		s.stopRemounter()
		s.resetState()
		s.state = stateMountRequested
		return "", err
	}

	return s.info.SessionId, nil
}

func (s *Session) fetchMountHeaders(ctx context.Context) {
	if val := ctx.Value(IdempotenceIdHeaderKey); val != nil {
		if str, ok := val.(string); ok {
			s.mountHeaders.IdempotenceId = str
		}
	}

	if val := ctx.Value(TraceIdHeaderKey); val != nil {
		if str, ok := val.(string); ok {
			s.mountHeaders.TraceId = str
		}
	}

	if val := ctx.Value(RequestTimeoutHeaderKey); val != nil {
		if d, ok := val.(time.Duration); ok {
			s.mountHeaders.RequestTimeout = uint32(durationToMsec(d))
		}
	}

	if val := ctx.Value(ClientIdHeaderKey); val != nil {
		if str, ok := val.(string); ok {
			s.mountHeaders.ClientId = str
		}
	}
}

func (s *Session) setupMountHeaders(ctx context.Context) context.Context {
	mountCtx := context.WithValue(
		ctx,
		IdempotenceIdHeaderKey,
		s.mountHeaders.IdempotenceId)

	mountCtx = context.WithValue(
		mountCtx,
		TraceIdHeaderKey,
		s.mountHeaders.TraceId)

	mountCtx = context.WithValue(
		mountCtx,
		RequestTimeoutHeaderKey,
		s.mountHeaders.RequestTimeout)

	mountCtx = context.WithValue(
		mountCtx,
		ClientIdHeaderKey,
		s.mountHeaders.ClientId)

	return mountCtx
}

func (s *Session) processMountResponse(
	ctx context.Context,
	diskId string,
	volume *protos.TVolume,
	info *SessionInfo,
	err error,
) error {

	if err == nil {
		if s.volume.BlockSize != 0 {
			// validate volume geometry
			if s.volume.BlockSize != volume.BlockSize {
				err = &ClientError{
					Code:    E_FAIL,
					Message: "Volume geometry changed",
				}
			}
		}
	}

	if err != nil {
		if logger := s.log.Logger(LOG_ERROR); logger != nil {
			logger.Printf(
				ctx,
				"MountVolume for %s failed: %v, client_id: %s",
				diskId,
				err,
				s.mountHeaders.ClientId,
			)
		}

		// we do not want to hide mount failure from client,
		// so request will NOT be automatically retried.
		// just force remount on next request
		s.state = stateMountRequested
		return err
	}

	if logger := s.log.Logger(LOG_DEBUG); logger != nil {
		logger.Printf(
			ctx,
			"Complete MountVolume for %s, client_id: %s",
			diskId,
			s.mountHeaders.ClientId,
		)
	}

	previousRemountPeriod := s.info.InactiveClientsTimeout

	s.volume = volume
	s.info = *info
	s.state = stateMountCompleted

	if s.info.InactiveClientsTimeout != previousRemountPeriod {
		s.stopRemounter()

		if s.info.InactiveClientsTimeout != 0 {
			s.startRemounter(ctx)
		}
	}

	return nil
}

func (s *Session) stopRemounter() {
	if s.remounter != nil {
		s.remounter.Done <- true
		close(s.remounter.Done)
		s.remounter.Ticker.Stop()
		s.remounter = nil
	}
}

func (s *Session) startRemounter(ctx context.Context) {
	s.remounter = &remounter{
		Ticker:     s.makeTicker(s.info.InactiveClientsTimeout),
		Done:       make(chan bool, 1),
		remountCtx: &remountCtx{ctx},
	}

	go func(r *remounter) {
		for {
			select {
			case <-r.Done:
				r.Ticker.Stop()
				return
			case <-r.Ticker.TickChan():
				s.mountLock.Lock()

				select {
				case <-r.Done:
				default:
					_, _ = s.remountVolume(r.remountCtx)
				}

				s.mountLock.Unlock()
				r.Ticker.TickProcessed()
			}
		}
	}(s.remounter)
}

////////////////////////////////////////////////////////////////////////////////

func NewSession(client Client, log Log) *Session {

	return &Session{
		client: client,
		log:    log,
		volume: new(protos.TVolume),
		makeTicker: func(period time.Duration) ticker {
			return newTimeTicker(period)
		},
	}
}
